### ISSUE TITLE
Incorrect shuffle optimization in LLVM12 (affecting ISPC 1.16.0)

### ISSUE THREAD

**Issue by dbabokin** [ispc/ispc#2111]
Original test case was reported by @JeffRous here: https://ispc.godbolt.org/z/7x4d8jnPq

This appears to be incorrect transformation in LLVM 12 (works fine with LLVM 11). Reduced test case is here:

```c
struct FVector3 {
    float x, y, z;
};

static const varying int vAOS32 = {0, 1, 2, 3, 8, 9, 10, 11};

export void test(uniform FVector3 &Out, uniform float *uniform In)
{
    varying float vx;

    const varying float src0 = In[programIndex];
    const varying float src1 = In[programCount + programIndex];

    vx = shuffle(src0, src1, vAOS32);

    Out.x = reduce_add(vx);
}
```

Compiling (`ispc --target=avx2-i32x8 fail_reduced.ispc --emit-asm`) with LLVM 11.1 (ISPC trunk) we get:
```asm
         vmovups (%rsi), %ymm0

         vperm2f128      $32, 32(%rsi), %ymm0, %ymm0 ## ymm0 = ymm0[0,1],mem[0,1]
         vhaddps %ymm0, %ymm0, %ymm0

         vhaddps %ymm0, %ymm0, %ymm0
         vextractf128    $1, %ymm0, %xmm1
         vaddss  %xmm1, %xmm0, %xmm0
         vmovss  %xmm0, (%rdi)
         vzeroupper
         retq
```

Same with LLVM 12.0 (ISPC trunk):
```asm
         vmovups (%rsi), %ymm0

         # This section is different and incorrect
         vhaddps 32(%rsi), %ymm0, %ymm0
         vpermpd $136, %ymm0, %ymm0              ## ymm0 = ymm0[0,2,0,2]

         vhaddps %ymm0, %ymm0, %ymm0
         vextractf128    $1, %ymm0, %xmm1
         vaddss  %xmm1, %xmm0, %xmm0
         vmovss  %xmm0, (%rdi)
         vzeroupper
         retq
```

**Comment by dbabokin** [comment]
To be more clear on the problem here. First sequence is doing the following:
```asm
         # ymm0 = [y0, y1, y2, y3, y4, y5, y6, y7]
         # mem = [z0, z1, z2, z3, z4, z5, z6, z7]
         vperm2f128      $32, 32(%rsi), %ymm0, %ymm0 ## ymm0 = ymm0[0,1],mem[0,1]

         # ymm0 = [y0, y1, y2, y3, z0, z1, z2, z3]
         vhaddps %ymm0, %ymm0, %ymm0

         # ymm0 = [y0+y1, y2+y3, y0+y1, y2+y3, z0+z1, z2+z3, z0+z1, z2+z3]
```

The second is doing this:
```asm
         # ymm0 = [y0, y1, y2, y3, y4, y5, y6, y7]
         # mem = [z0, z1, z2, z3, z4, z5, z6, z7]
         vhaddps 32(%rsi), %ymm0, %ymm0

         # ymm0 = [z0+z1, z2+z3, y0+y1, y2+y3, z5+z6, z7+z8, y5+y6, y7+y8]
         vpermpd $136, %ymm0, %ymm0              ## ymm0 = ymm0[0,2,0,2] <<<<<< note this!
         # It should be ymm0[1,1,0,0] to be equivalent transformation.

         # ymm0 = [z0+z1, z2+z3, z5+z6, z7+z8, z0+z1, z2+z3, z5+z6, z7+z8]
```

**Comment by DeepakRajendrakumaran** [comment]
https://bugs.llvm.org/show_bug.cgi?id=50823#c3 : corresponding LLVM bug

**Comment by dbabokin** [comment]
This bug was fixed in LLVM trunk, but was too late to be merged to 12.0.1, @DeepakRajendrakumaran could you please create a regression test and add a patch to ISPC repo for 12.0?
